Changelog authoring guide (#1727)
hashicorp/design-system: wiki/Website-Changelog.md

# Changelog authoring best practices

- [Tips for writing consistent changelog entries](#tips-for-writing-consistent-changelog-entries)
  - [General guidance](#general-guidance)
  - [Consistent terminology](#consistent-terminology)
  - [Component references](#component-references)
  - [Content formatting](#content-formatting)
    - [Component changes](#component-changes)
    - [New icons](#new-icons)
- [Templates](#templates)
  - [Templates for npm packages](#templates-for-npm-packages)
  - [Templates for Figma changelog entries](#templates-for-figma-changelog-entries)

---

## Tips for writing consistent changelog entries

### General guidance
- Changelog entries should always start with a paragraph (not a list).
- Consider using a list instead of a long paragraph of text to communicate multiple changes for one element.
- If you need to include a note (like a migration tip), add it to the bottom of the entry in a new paragraph. Consider using emphasized text if you need to draw more attention to it.

- Entries should use complete sentences but be short, clear, and descriptive enough to be helpful.
- Use past tense.
- Entries should always end with a period.

> 

Reviewer: ```suggestion
- If there are changes to multiple components in the same entry, each change should be described in its own paragraph (and list of details, if needed)

```

### Consistent terminology
- Use consistent terms for different types of changes: 
  - Bugfix: “Fixed”
  - New component, token, variant: “Added”
  - Update: “Changed”, “Refactored”
  - Removed or deleted: “Removed”

>   - Other edge cases: “Upgraded”, “Reduced”, “Prevented”

Reviewer: ```suggestion
  - Other possible cases: “Upgraded”, “Reduced”, “Prevented”
```

 
### Component references
- When referencing components, use the plain-text component name, e.g., `Accordion`.
- All components or token names, including those in the summary, should be enclosed in backticks (`).

### Content formatting

While best practice guidelines are provided, clarity is most important, so use your best judgment. 

#### Component changes

For component changes, place the component name first and enclose it in backticks (`), followed by a dash (-), and then a brief summary.

| Do | Dont |
|:---|:-----|
| `Dropdown` - changed the `@height` property to use `max-height` instead of a fixed height.  | Changed the `@height` property in `Dropdown` to use `max-height` instead of a fixed height. |

#### New icons

For icons, use a comma-separated list instead of a bulleted list and start with the icon values. Icon names should be enclosed in backticks (`) and use the Ember value.

| Do | Dont |
|:---|:-----|

> | `twitter-x`, `twitter-x-color`, `aws-cdk`, `aws-cdk-color`, `jfrog`, and `jfrog-color` icons added. | Added a new set of service icons: <ul><li>`twitter-x`</li><li>`twitter-x-color`</li><li>`aws-cdk`</li><li>`aws-cdk-color`</li><li>`jfrog`</li><li>`jfrog-color`</li></ul> | 

Reviewer 2: I think keeping the icons in a sentence format works well; for example, if we have more than three icons, the list can get long and take up more space than needed.

---

## Templates

### Templates for npm packages

Copy and paste this template and adjust as necessary when creating a new [changeset](https://github.com/hashicorp/design-system#changesets). _Replace all elements in brackets._

#### Simple entry
```
`{component-name}` - Fixed {...additional details}.
```

#### Multiple changes per element
```
`{component-name}` - Changed {...additional details}, including:
- {change 1}
- {change 2}
- etc

_{additional details}..._
```

> 

Reviewer: ````suggestion

#### Multiple elements changed
```
`{component-#1-name}` - Changed {...additional details}, including:
- {change 1}
- {change 2}

`{component-#2-name}` - Refactored {...additional details}

_{...important details}_
```

````

# Changelog authoring best practices

- [Tips for writing consistent changelog entries](#tips-for-writing-consistent-changelog-entries)
- [Content structure](#content-structure)
  - [Component changes](#component-changes)
  - [New icons](#new-icons)
  - [Breaking changes](#breaking-changes)
- [Templates](#templates)
  - [Templates for npm packages](#templates-for-npm-packages)
  - [Templates for Figma changelog entries](#templates-for-figma-changelog-entries)

---

## Tips for writing consistent changelog entries

> - Changelog entries should always start with a paragraph (not a list).
> - Consider using a list instead of a long paragraph of text to communicate multiple changes for one element.
> - If you need to include a note (like a migration tip), add it to the bottom of the entry in a new paragraph. Consider using emphasized text, if you need to draw more attention to it.
> - Entries should use complete sentences but be short, clear, and descriptive enough to be helpful.
> - Use past tense.
> - Entries should always end with a period.
> - Use consistent terms for different types of changes: 
>   - Bugfix: “Fixed”
>   - New component, token, variant: “Added”
>   - Update: “Changed”, “Refactored”
>   - Removed or deleted: “Removed”
>   - Other edge cases: “Upgraded”, “Reduced”, “Prevented”
> - When referencing components, use the plain-text component name, e.g., `Accordion`.
> - All components or token names, including those in the summary, should be enclosed in backticks (`).

Reviewer 2: ```suggestion
**General Guidelines:**
- Start changelog entries with a paragraph, not a list.
- If multiple changes are made to a single element, consider listing them instead of creating a long text paragraph.
- If there's a need to include a note (e.g., a migration tip), place it at the bottom of the entry in a new paragraph. Emphasize the text to draw more attention if necessary.

**Writing Style:**
- Use complete sentences that are short, clear, and descriptive.
- Write in the past tense.
- End entries with a period.

**Terminology:**
- Utilize consistent terms for different types of changes:
  - Bugfix: "Fixed"
  - New additions: "Added" (e.g., new component, token, variant)
  - Updates: "Changed", "Refactored"
  - Removals: "Removed"
  - Other situations: "Upgraded," "Reduced," "Prevented."

**Component Referencing:**
- Reference components using their plain-text name, like `Accordion.`
- Enclose all component or token names, including those in the summary, in backticks (`).
```

Reviewer 2: Slight suggestion for breaking up the content here

Author: @andgen404 I didn't make the change exactly as you proposed, but I made adjustments to use headings to break up the content better. Thanks for the suggestion!

## Content structure

While best practice guidelines are provided, clarity is most important, so use your best judgment. 

### Component changes

For component changes, place the component name first and enclose it in backticks (`), followed by a dash (-), and then a brief summary.

| Do | Dont |
|:---|:-----|
| `Dropdown` - changed the `@height` property to use `max-height` instead of a fixed height.  | Changed the `@height` property in `Dropdown` to use `max-height` instead of a fixed height. |

### New icons

> For icons, use a comma-separated list instead of a bulleted list and start with the icon values. Icon names should be enclosed in backticks (`) and use the Ember value.

Reviewer: I've personal preference for a list, but not a blocker. Just a note.

Author: I had that at first, but some of the lists got pretty long, so I changed it. I don't have strong opinions one way or the other, so curious what others think.

| Do | Dont |
|:---|:-----|
| `twitter-x`, `twitter-x-color`, `aws-cdk`, `aws-cdk-color`, `jfrog`, and `jfrog-color` icons added. | Added a new set of service icons: <ul><li>`twitter-x`</li><li>`twitter-x-color`</li><li>`aws-cdk`</li><li>`aws-cdk-color`</li><li>`jfrog`</li><li>`jfrog-color`</li></ul> | 

### Breaking changes

For breaking changes, add “⛔️ **Breaking change** - ” after the component name/before the summary. Be sure to include relevant details about why this is a breaking change and what needs to be done on the consumer side in order to upgrade to this version.

| Do | Dont |
|:---|:-----|

> | **`Dropdown ListItem`** - ⛔️ **Breaking change** - Renamed internal CSS classes as follows: <ul><li>from `hds-dropdown-list-item–copy-item` to `hds-dropdown-list-item–variant-copy-item`</li><li>from `hds-dropdown-list-item–description` to `hds-dropdown-list-item–variant-description`</li><li>from `hds-dropdown-list-item–generic` to `hds-dropdown-list-item–variant-generic`</li><li>from `hds-dropdown-list-item–interactive` to `hds-dropdown-list-item–variant-interactive`</li><li>from `hds-dropdown-list-item–separator` to `hds-dropdown-list-item–variant-separator`</li><li>from `hds-dropdown-list-item–title` to `hds-dropdown-list-item–variant-title`</li></ul><br/>_If test assertions are relying on these class names, tests will fail. If extensions/overrides have been applied to these classes, they will suffer visual changes._ | Renamed `Hds::Dropdown::ListItem` internal CSS classes. |

Reviewer: In this example, the component name is bold, while in the other examples it's not. I think we should be consistent.

For context: I tried to set the component name bold, in one of my tests, but the result was not good: it added noise, without improving visual scanning. So my preference goes to using regular text everywhere, and leave bold text for when the author wants to highlight something.

Author: Ahh, the initial intention was to have them always be bold (I think I missed one example), but after comparing in your other branch, I agree that it makes them a bit too heavy, so I'll adjust the guidance.

---

## Templates

### Templates for npm packages

Copy and paste this template and adjust as necessary when creating a new [changeset](https://github.com/hashicorp/design-system#changesets). _Replace all elements in brackets._

#### Simple entry
```

> **`{component-name}`** - Fixed {...additional details}.

Reviewer: Ok, here's is bold again. Now I am not sure anymore which one you prefer. Do you want me to do a test for you where I show you the before and after?

_{additional details}..._
```

#### Breaking changes
```
**`{component-name}`** - ⛔️ **Breaking change** - Changed {...additional details}.

_{additional details}..._
```

<details>
  <summary><h4>Example outputs for npm packages</h4></summary>
  <i>These examples are just outputs of the details provided as a part of the changeset. They do not include the PR links or contributors, which are added automatically.</i>
  <br/><br/>
  
  **`TooltipButton`** - Added `text-align: inherit` to the “button” element.
  
  **`Dropdown`** - Fixed a few accessibility failures, including:
  - Added `aria-hidden` to the separator,
  - Added `role=“none”` to the checkbox `<li>`.
    
  **`Table`** - Changed `height` to `min-height` for the table head cells and updated the cells' internal padding to align with the design specs in Figma.
  
  **`Sidenav Link`** - Fixed an issue resulting in an empty node.
  - This will lead to a minimal visual impact on some edge cases of `Alert` and `Toast` (multiple description items) and `Sidenav` (text and generic content).
  
  **`Standalone Link`** - Increased the space between the actions to 16px and reduced the size of the icon in the small variant to 12px.

>   **`Modal`** and **`Flyout`** - Fixed the scroll management resulting in stray `style` attribute on the `<body>` element.

Reviewer: ```suggestion

  **`Modal`** and **`Flyout`** - Fixed the scroll management resulting in stray `style` attribute on the `<body>` element.
```

  Updated the CSS of multiple components to use flex `gap`.
  Upgraded Ember.js to the latest stable release 4.12, including upgrades to:
  - `ember-auto-import` from `2.6.0` to `2.6.3`
  - `ember-cli-htmlbars` from `6.1.0` to `6.2.0`
  
  Upgraded the following dependencies:
  - `ember-focus-trap` from `1.0.1` to `1.0.2`
  - `ember-keyboard` from `8.1.0` to `8.2.0`
  - `ember-truth-helpers` from `3.0.0` to `3.1.1`
  - `sass` from `1.58.3` to `1.62.1`
  
  Shifted our support version of Node.js from `12.* || 14.* || >= 16` to `14.* || 16.* || >= 18`
</details>

### Templates for Figma changelog entries

> Entries made for the Figma Components and Foundations UI Kits require a few additional steps. Copy and paste this template and adjust as necessary after publishing changes in Figma. _Replace all elements in brackets._

Reviewer 2: ```suggestion
Entries for the Figma Components and Foundations UI Kits have a few additional steps: Copy and paste this template and make necessary adjustments after publishing the changes in Figma. Utilize abbreviated formats for months. Replace all elements in brackets.``

#### Simple entry
```
### {month day, year}

**`{component-name}`** - Fixed {...additional details}.
```

#### Multiple changes per element
```
### {month day, year}

**`{component-name}`** - Changed {...additional details}, including:
- {change 1}
- {change 2}
- etc

_{additional details}..._
```

#### Breaking changes
```
### {month day, year}

**`{component-name}`** - ⛔️ **Breaking change** - Changed {...additional details}.

_{additional details}..._
```

<details>
  <summary><h4>Example outputs for Figma changes</h4></summary>

>   <h3>September 15, 2023</h3>

Reviewer 2: ```suggestion
  <h3>Sep 15, 2023</h3>
```

Reviewer 2: Could we consider abbreviating the month names, similar to the format used in the Changelog in Figma? This change could help create a consistent visual pattern within the sidecar.

Author: I agree that there should be a consistent pattern, but I disagree with it being the abbreviated version. Not all screen readers will interpret the abbreviated version as the month, instead reading it out at "Sep". Here are a few examples of how dates are read out to some screen readers: https://pauljadam.com/demos/dates.html
